stdlib/public/core/Array.swift

```diff
@@ -1865,3 +1865,187 @@ internal struct _ArrayAnyHashableBox<Element: Hashable>
     return true
   }
 }
+
+// SWIFT_ENABLE_TENSORFLOW
+// `Array<Element>` is a countably infinite direct sum 
+// (https://en.wikipedia.org/wiki/Direct_sum) of `Element`, when `Element` 
+// conforms to `AdditiveArithmetic` and thus has a zero.
+//
+// Therefore, the most useful (co)tangent space for `Array<Element>` is the 
+// countably infinite direct sum of 
+// `Element.TangentVector`/`Element.CoTangentVector` (since 
+// `Element.TangentVector` and `Element.CoTangentVector` conform to 
+// `AdditiveArithmetic`, they have a zero). If we think of all the elements past 
+// the end of an array as zeros, values of these direct sums can be represented 
+// by values of `Array<Element.TangentVector>`/`Array<Element.CotangentVector>`. 
+// 
+// The operations are:
+// - `zero` is `[]` (which is equivalent to `[.zero]`, `[.zero, .zero]`, etc).
+// - `+` is (pseudocode): 
+//   ```
+//   a + b = range(max(a.count, b.count)).map {
+//     i in (a[i] or .zero) + (b[i] or .zero)
+//   }
+//   ```
+// 
+// This leads to expected behavior when differentiating functions on arrays. 
+// For example:
+// ```
+// func sumFirstThree(_ array: Array<Float>) -> Float {
+//   return array[0] + array[1] + array[2]
+// }
+// ```
+// We expect the gradient of that to be `[1, 1, 1]`. That is what we get, if we 
+// define the pullback of array subscripting as follows:
+// ```
+// func vjpSubscript(index) {
+//   let result = self[index]
+//   let pullback = { v in
+//     var pb = Array(repeating: zero, count: index + 1)
+//     pb[index] = v
+//     return pb
+//   }
+//   return (result, pullback)
+// }
+// ```
+//
+extension Array : Differentiable where Element : Differentiable {
+  public typealias AllDifferentiableVariables = [Element]
```

Comment:
I think that with some extra work it would be possible to make this type be `TangentVector` when `Element.TangentVector == Element`. (And similarly for `CotangentVector`). This would make Arrays compatible with all the optimizers in swift-apis that require `Model.CotangentVector == Model.AllDifferentiableVariables`!

No need to do this in this PR though.

Reply:
Multiple conformances to the same protocol aren't possible, however.

We could work around it by introducing a new protocol refining `Differentiable` with additional constraints `TangentVector == CotangentVector == AllDifferentiableVariables`. Key-path-based optimizers would then constrain `Model` to this new protocol.

Future work for sure.

Reply:
Oh interesting. I wasn't aware of that constraint in the optimizers. Would the type checker not be able to prove this automatically when `Element.TangentVector == Element`? If not, how can I add support for this? I noticed that if I simply add other more specialized extensions,  I cannot redeclare the type alias.

Reply:
So in the current form the optimizers would not be able to work with differentiable arrays, right?

Reply:
> So in the current form the optimizers would not be able to work with differentiable arrays, right?

Yep. Except for the riemann sgd one. That doesn't have any constraints, and it uses `moved`.

I think it won't be too hard to get this working with optimizers. Things you'll need to do:
* Add a constraints to this conformance that `Element.AllDifferentiableVariables == Element.TangentVector` and `Element.AllDifferentiableVariables == Element.CotangentVector`. (Now Arrays with Elements not satisfying this constraint won't be differentiable, and there's no way to work around this because of the multiple-conformance thing that @dan-zheng mentioned. But pretty much every practical differentiable type satisfies these constraints, so I don't think this will be very bad.)
* Define only a single struct, and set `AllDifferentiableVariables`, `TangentVector`, and `CotangentVector` to be that struct.
* There might be some issue where the optimizer tries to access out of bounds indices because the gradient ends up with a smaller array from the original or vice-versa. Add something that handles that by returning the right zeros.

Reply:
Ok, I'll got ahead and do that because I feel this is the main use of the `Differentiable` protocol and given the current multiple-conformance limitation, this is better than the alternative of having differentiable arrays that cannot be used with optimizers.

Regarding your last point, I don't believe the gradient can end up with a smaller array from the original given that we currently never shrink the array if it contains `.zero`. We only expand them.

Reply:
IMO we should not add a `AllDifferentiableVariables == CotangentVector` constraint -- it's inconsistent with the general rules that we already derive in the compiler. The current implementation is **almost** there, and I don't think there's any practical type system constraints at all.

To start, the following associated type is incorrect. 
```swift
public typealias AllDifferentiableVariables = [Element]
```
It should be:
```swift
public typealias AllDifferentiableVariables = [Element.AllDifferentiableVariables]
```

If you change this, other errors will likely pop up and point you to the right direction.

Reply:
So I'm working on this right now and I'm not sure how to implement the setter for `allDifferentiableVariables`. I want to apply the setter on each element of `self`, but the elements of `self` are immutable. So, the following does not work:
```
  public var allDifferentiableVariables: AllDifferentiableVariables {
    get {
      return self.map { $0.allDifferentiableVariables }
    }
    set {
      zip(self, newValue).forEach { $0.allDifferentiableVariables = $1 }
    }
  }
```

Reply:
Indexing from the _root_ mutable variable is currently the only way to get an `inout` reference.
```swift
set {
  for i in indices {
    self[i].allDifferentiableVariables = newValue[i]
  }
}
```

Reply:
Good to know that! Thanks! :) I'm still getting familiar with the language.

Reply:
So, currently I can only make it work if I add the extra condition that `Element.AllDifferentiableVariables == Element`. Otherwise, I can't satisfy the `'Self.CotangentVector' == 'Self.AllDifferentiableVariables.CotangentVector'` requirement.

Reply:
`AllDifferentiableVariables` should be a separate nested struct.

Reply:
Right, I think this resolved it. All that remains is to address Marc's comment about `moved` and `tangentVector`.

Reply:
Awesome!